### PR TITLE
Add support for pulp for RPM repository management

### PR DESCRIPTION
**Pull request by cottsay**
Here are some examples of what the corresponding config changes look like:

`hiera/hieradata/buildfarm_role/master.yaml`:
```diff
@@ -5,6 +5,10 @@ credentials::jenkins-slave::username: jenkins-agent
 credentials::jenkins-slave::id: 1e7d4696-7fd4-4bc6-8c87-ebc7b6ce16e5
 credentials::jenkins-slave::passphrase: 4lRsx/NwfEndwUlcWOOnYg==
 
+credentials::pulp-api::username: admin
+credentials::pulp-api::id: 0ddb9953-40d3-4922-80be-298086e30370
+credentials::pulp-api::passphrase: asdf1234
+
 # Uncomment below to increase the memory available for Jenkins
 # This is highly recommended for larger deployments
 # Don't use more than 75% of your RAM for Jenkins
```

`hiera/hieradata/buildfarm_role/repo.yaml`:
```diff
@@ -131,3 +131,12 @@ jenkins-agent::reprepro_config:
             # needed until we deploy the gpg key
             verify_release: blindtrust
 
+jenkins-agent::pulp_config:
+    admin_passphrase: asdf1234
+    rpm:
+        el:
+            architectures: [aarch64, x86_64]
+            versions: ['7', '8']
+        fedora:
+            architectures: [aarch64, armhf, i386, x86_64]
+            versions: ['30', '31']
```

The approach is to create a CentOS 8 container that installs and provisions pulpcore, and run that container using a systemd service. There is also a python script for creating the initial repositories and endpoints in pulp once it is up-and-running.

I'm pretty new to Puppet, Hiera and Ruby. If something looks wonky, it's probably a mistake, so let me know.

### CODE REVIEW

**Comment by nuclearsandwich** [comment]
> The approach is to create a CentOS 7 container that installs and provisions pulpcore, and run that container using a systemd service

Does pulp itself need to run on CentOS in order to serve RPMs?

**Comment by cottsay** [comment]
> Does pulp itself need to run on CentOS in order to serve RPMs?

It might not. At first, I firmly believed that it did, but after many iterations I realized that the python module for createrepo was re-building createrepo from source as part of the pip package (the version of createrepo in Ubuntu is insufficient).

*Maybe* we can run pulp on Xenial. I'll give it a shot, at least.

**Comment by cottsay** [comment]
After digging into this, it doesn't look like running pulp natively on Ubuntu is a good idea right now. There are missing dependencies that would need to be built from source, and some dependencies that might be too old.

I changed this PR to move the non-pulp components (psql and redis) out of the container, simplified the Dockerfile, and split out a service for each of the pulp components.

**Comment by cottsay** [comment]
So I don't think my `credentials::pulp-api::password_secret` trick is working. If I understand correctly, the Jenkins secrets are salted in some way, and aren't portable values.

It looks like the Jenkins puppet module actually talks to the Jenkins API to add credentials.

@nuclearsandwich, do you have any insights here?

**Comment by nuclearsandwich** [comment]
> So I don't think my `credentials::pulp-api::password_secret` trick is working. If I understand correctly, the Jenkins secrets are salted in some way, and aren't portable values.

Jenkins secrets are AES encrypted with a master key generated for each deployment. I was under the impression that unencrypted secrets would be noticed and encrypted when the config was read and written back out but maybe that's not correct. I'll look into it and get back to you.

**Review by nuclearsandwich — COMMENTED** [review]
I've so far reviewed for puppet organization and one or two individual items that jumped out at me.

One general comment is that I notice a lot of exec resources that include `test || set` and you could separate the test into an `onlyif` or `unless` parameter. [docs](https://puppet.com/docs/puppet/3.8/types/exec.html)

File: modules/pulp/manifests/init.pp [hunk 1]
```diff
@@ -0,0 +1,13 @@
+class pulp {
```
[the remark this answers is not in the document]
Comment (cottsay) [reply modules/pulp/manifests/init.pp]: This was completely out of ignorance. I refactored it to look more like the mainstream modules, and the `core` functionality was moved to `init`.

File: modules/profile/manifests/ros/rpm_repo.pp [hunk 2]
```diff
@@ -0,0 +1,65 @@
+define profile::ros::rpm_repo(
```
[the remark this answers is not in the document]
Comment (cottsay) [reply modules/profile/manifests/ros/rpm_repo.pp:1]: This `rpm_repo` stuff represents a single repository instance. Even a single RPM repo/arch will have at least 3 of these. My understanding is that a `class` would not be appropriate.

This resource is only referenced in one spot, so I could just copy/paste the code there if you like.

File: modules/pulp/manifests/core.pp [hunk 3]
```diff
@@ -0,0 +1,228 @@
...
+    enable => true,
+    require => Package['redis-server'],
+  }
+
+  exec { 'redis_config':
+    path => ['/bin', '/usr/bin'],
+    command => "sed -i -E -e 's|^#? ?port .*|port 0|' -e 's|^#? ?unixsocket .*|unixsocket /var/run/redis/redis.sock|' -e 's|^#? ?unixsocketperm .*|unixsocketperm 777|' /etc/redis/redis.conf",
+    require => Package['redis-server'],
+    notify => Service['redis'],
+  }
```
[the remark this answers is not in the document]
Comment (cottsay) [reply modules/pulp/manifests/core.pp]: The `sed` really isn't that bad, it's just 3 simple operations in a single invocation. I broke it up into separate lines.

We *could* dictate the entire redis config, but I don't really think it's necessary. This sed is idempotent and only modifies the parts of the config we actually require.

File: modules/pulp/templates/pulp_sign.sh.epp [hunk 4]
```diff
@@ -0,0 +1,32 @@
+#!/usr/bin/env bash
+
+export GNUPGHOME=/home/pulp/.gnupg
+
+FILE_PATH=$1
+SIGNATURE_PATH="$1.asc"
+
+PUBLIC_KEY_PATH="$(cd "$(dirname $1)" && pwd)/public.key"
+
+ADMIN_ID="$gpg_key_id"
```
Comment (nuclearsandwich) [inline modules/pulp/templates/pulp_sign.sh.epp:10]: The unescaped dollars give me slight pause. To make the difference between template and shell variables more explicit we could turn this into an ERB file and use:

```suggestion
ADMIN_ID="<%= gpg_key_id %>"
```

That might need to be `@gpg_key_id`, I need to refresh how puppet sets up erb rendering.

**Review by nuclearsandwich — COMMENTED** [review]
Looks good so far, some minor comments from review session.

**Comment by ros-discourse** [comment]
This pull request has been mentioned on **ROS Discourse**. There might be relevant details there:

https://discourse.ros.org/t/rpm-packaging-support-in-ros-buildfarm/15672/1

**Pull request closed without merging**